lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -135,6 +135,151 @@ Value createLLVMIntegerConstant(OpBuilder &builder, Location loc, short width,
   LLVM::createIndexConstant(rewriter, loc, this->getTypeConverter(),           \
                             __VA_ARGS__)
 
+LLVM::LLVMFuncOp getVprintfDeclaration(OpBuilder &builder) {
+  auto moduleOp = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  StringRef funcName("vprintf");
+  Operation *funcOp = moduleOp.lookupSymbol(funcName);
+  if (funcOp)
+    return cast<LLVMFuncOp>(*funcOp);
+
+  auto *context = builder.getContext();
+  auto retType = builder.getI32Type();
+
+  SmallVector<Type> argsType{
+      LLVM::LLVMPointerType::get(IntegerType::get(context, 8)),
+      LLVM::LLVMPointerType::get(IntegerType::get(context, 8))};
+  auto funcType = LLVM::LLVMFunctionType::get(retType, argsType);
+
+  ConversionPatternRewriter::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(moduleOp.getBody());
+
+  return builder.create<LLVMFuncOp>(UnknownLoc::get(context), funcName,
+                                    funcType);
+}
+
+std::pair<Type, Value> promoteValue(OpBuilder &builder, Value value) {
```

Comment:
You can probably include `getVprintfDeclaration`, `promoteValue`, and `llPrintf` into `PrintfOpConversion`. I think they won't be used by external classes.

Reply:
let me do this.